src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -346,44 +346,90 @@ private static void GetHexStringCore(Span<char> destination, bool lowercase)
 
-                while (!destination.IsEmpty)
+                if (BitOperations.IsPow2(choices.Length))
                 {
-                    if (destination.Length < randomBytes.Length)
+                    // To avoid bias, we can't just % all bytes to get them into range; that would cause
+                    // the lower values to be more likely than the higher values. If the number of choices
+                    // is a power of 2, though, we can just mask off the extraneous bits.
+
+                    int mask = choices.Length - 1;
+
+                    while (!destination.IsEmpty)
                     {
-                        randomBytes = randomBytes.Slice(0, destination.Length);
+                        // If this will be the last iteration, avoid over-requesting randomness.
+                        if (destination.Length < randomBytes.Length)
+                        {
+                            randomBytes = randomBytes.Slice(0, destination.Length);
+                        }
+
+                        RandomNumberGeneratorImplementation.FillSpan(randomBytes);
+
+                        for (int i = 0; i < randomBytes.Length; i++)
+                        {
+                            destination[i] = choices[randomBytes[i] & mask];
+                        }
+
+                        destination = destination.Slice(randomBytes.Length);
                     }
+                }
+                else
+                {
+                    // As the length isn't a power of two, we can't just mask off all extraneous bits, and
+                    // instead need to do rejection sampling. However, we can mask off the irrelevant bits, which
+                    // then reduces the chances of needing to reject a value.
 
-                    RandomNumberGeneratorImplementation.FillSpan(randomBytes);
+                    int mask = (int)BitOperations.RoundUpToPowerOf2((uint)choices.Length) - 1;
 
-                    int mask = choices.Length - 1;
-                    for (int i = 0; i < randomBytes.Length; i++)
+                    while (!destination.IsEmpty)
                     {
-                        destination[i] = choices[randomBytes[i] & mask];
+                        // Unlike in the IsPow2 case, where every byte will be used, some bytes here may
+                        // be rejected. On average, half the bytes may be rejected, so we heuristically
+                        // choose to shrink to twice the destination length.
+                        if (destination.Length * 2 < randomBytes.Length)
+                        {
+                            randomBytes = randomBytes.Slice(0, destination.Length * 2);
+                        }
+
+                        RandomNumberGeneratorImplementation.FillSpan(randomBytes);
+
+                        int i = 0;
+                        foreach (byte b in randomBytes)
+                        {
+                            if ((uint)i >= (uint)destination.Length)
```

Comment:
`i` can only meet/exceed destination.Length when incremented.  If this `if` is moved to within the "non-rejected" case it would remove a jump statement from the rejected sample iterations.

It might have no measurable effect, given that the loop body is fairly short.  But at 100_000 destination values with a choice-set of 192 (halfway between 128 and 256) you might see something.

Reply:
That comment was based on the assumption that `destination.Length == 0` was checked for in the callers.  It doesn't seem to be true.  So moving the `if` deeper would require adding those checks either at the beginning of this method or into the callers.

Reply:
I can test it, but I expect there's going to be a branch anyway as part of the bounds check when indexing into destination. This check here should enable the JIT to remove that bounds check, so it should in theory come out in the wash.